---
title: "Visualising the phylogenetic distribution of bacterial human pathogens."
author: "Daniel Padfield"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: 
  html_document:
    code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, tidy.opts=list(width.cutoff=50), tidy = TRUE, warning = FALSE)
```

## Dataset of bacterial human pathogens

We present a comprehensive list of known bacterial pathogens known to cause infectious symptoms in humans. A summary of all the datasets used and produced are summarised in this document.

The dataset is available on [GitHub](https://github.com/padpadpadpad/bartlett_et_al_2022_human_pathogens/tree/master/data).

This work-through recreates Figure 4 and Table 2 of the manuscript. It shows how to we linked the GTDB and NCBI taxonomies of our pathogen list and how we plotted these on a phylogenetic tree of bacteria produced by GTDB.

### Load in packages and data

First we will load in the R packages used in the script.

```{r package_setup, message=FALSE}
# load packages
library(tidyverse)
library(ggtree)
library(ape)
library(RColorBrewer)
library(ggnewscale)
library(gt)
```

### Load in data from GTDB

We will use the bacterial phylogeny produced by [GTDB](https://gtdb.ecogenomic.org/). The Genome Taxonomy Database (GTDB) is an initiative to establish a standardised microbial taxonomy based on genome phylogeny, with the bacterial taxonomy based on genome trees inferred using FastTree from an aligned concatenated set of 120 single copy marker proteins.

We can load in the phylogeny from the most recent release at the time this project was done (r207).

```{r load_tree}
# read in tree
tree <- read.tree('https://data.gtdb.ecogenomic.org/releases/release207/207.0/bac120_r207.tree')
tree
```
Each tip of the tree is an accession number of the genome included in GTDB. We need a way to link these accession numbers to the species in our pathogen list. However, our taxonomy is assigned using NCBI nomenclature and the taxonomy from GTDB is - well - GTDB taxonomy. 

With each release GTDB have a file called `bac120_metadata_r207.tar.gz` where r207 represented the release of the database which contains both gtdb and ncbi taxonomic information of each genome. We will use this file to create a dataframe with both ncbi and gtdb taxonomy for each genome in GTDB. 

```{r gtdb_data, eval=FALSE}
# metadata file - its a big file
url <- 'https://data.gtdb.ecogenomic.org/releases/release207/207.0/bac120_metadata_r207.tar.gz'
download.file(url, destfile="tmp.tar.gz")

# unzip file
untar("tmp.tar.gz")

# read in gtdb taxonomy
taxonomy <- read_tsv('bac120_metadata_r207.tsv', col_names = TRUE) %>%
  # rename columns
  select(accession, gtdb_taxonomy, ncbi_taxonomy) 
```

```{r gtdb_data2, echo=FALSE}
# read in gtdb taxonomy
taxonomy <- read_tsv('bac120_metadata_r207.tsv', col_names = TRUE) %>%
  # rename columns
  select(accession, gtdb_taxonomy, ncbi_taxonomy) 
```

We can then remove all entries that are not present in the phylogenetic tree and split the taxonomies into each phylogenetic level. Finally we will keep only one entry of each NCBI species in GTDB database to make it easier to assign each member of our pathogen list a GTDB accession number.

```{r gtdb_clean}
taxonomy <- filter(taxonomy, accession %in% tree$tip.label) %>%
  # split taxonomy into columns for each level
  separate(gtdb_taxonomy, c('kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species'), sep = ';') %>%
  # get rid of p__ c__ etc
  mutate(across(kingdom:species, function(x){gsub('.*__', '', x)})) %>%
  separate(ncbi_taxonomy, c('ncbi_kingdom', 'ncbi_phylum', 'ncbi_class', 'ncbi_order', 'ncbi_family', 'ncbi_genus', 'ncbi_species'), sep = ';') %>%
  mutate(across(ncbi_kingdom:ncbi_species, function(x){gsub('.*__', '', x)}))

# keep only single member of each NCBI species in the taxonomy
ncbi_taxonomy <- taxonomy %>%
  group_by(ncbi_species) %>%
  slice_head(n = 1) %>%
  ungroup()

```


### Load in pathogen list

We can read in the pathogen list and rename the taxonomy columns to be the same as the ones in the GTDB taxonomy. We also need to concatenate genus and species together for the species column.

```{r load_pathogen_list}
# set url
url <- "https://github.com/padpadpadpad/bartlett_et_al_2022_human_pathogens/raw/master/data/bacteria_human_pathogens.xlsx"

d_pathogens <- rio::import(url, sheet = 'Tab 6 Full List', range = 'A1:J1514') %>%
  janitor::clean_names() %>%
  select(superkingdom:species, year, status) %>%
  rename_with(., function(x){paste('ncbi_', x, sep = '')}, superkingdom:species) %>%
  rename(ncbi_kingdom = ncbi_superkingdom) %>%
  mutate(ncbi_species = paste(ncbi_genus, ncbi_species, sep = ' '))

colnames(d_pathogens)

```

From trying to link GTDB NCBI taxonomy to the NCBI taxonomy assigned by [taxize](https://docs.ropensci.org/taxize/index.html), we noticed some minor naming discrepancies in species names that we corrected in our pathogen dataset just to allow easy linking between datasets. These changes are not maintained anywhere and are just to allow the tree to be plotted effectively.

```{r name_changes}
# make some specific changes to ncbi species based on mismatches
d_pathogens <- mutate(d_pathogens, ncbi_species = ifelse(ncbi_species == 'Neoehrlichia mikurensis', 'Candidatus Neoehrlichia mikurensis', ncbi_species),
                      ncbi_species = ifelse(ncbi_species == 'Catabacter hongkongensis', 'Christensenella hongkongensis', ncbi_species),
                      ncbi_species = ifelse(ncbi_species == 'Chlamydophila psittaci', 'Chlamydia psittaci', ncbi_species),
                      ncbi_species = ifelse(ncbi_species == 'Chlamydophila pneumoniae', 'Chlamydia pneumoniae', ncbi_species),
                      ncbi_species = ifelse(ncbi_species == 'Faecalicatena Contorta contorta', 'Faecalicatena contorta', ncbi_species),
                      ncbi_species = ifelse(ncbi_species == 'Metamycoplasma hominis', 'Mycoplasma hominis', ncbi_species),
                      ncbi_species = ifelse(ncbi_species == 'Metamycoplasma salivarium', 'Mycoplasma salivarium', ncbi_species),
                      ncbi_species = ifelse(ncbi_species == 'Mycoplasmoides genitalium', 'Mycoplasma genitalium', ncbi_species),
                      ncbi_species = ifelse(ncbi_species == 'Mycoplasmoides pneumoniae', 'Mycoplasma pneumoniae', ncbi_species))

```

### Linking GTDB taxonomy with pathogen list

We can link the GTDB taxonomy to our pathogen list (based on NCBI taxonomy). The aim of linking these dataframes is to give each member of our pathogen list a tip.label on the phylogenetic tree so we can visualise the distribution of pathogens on the phylogenetic tree. 

Ideally each NCBI species name in our pathogen list would be present in the NCBI species names of an accession of the GTDB database. However, as we are interested in visualising the tree at higher taxonomic levels (Order), we can assign each species in our list to an accession number of something in the same genus (or family if there are also no hits there).

```{r link_dataframes}
# add in gtdb taxonomy in stages

# first we will check for exact species matches
d_pathogen_species <- left_join(d_pathogens, select(ncbi_taxonomy, accession:species, ncbi_species)) %>%
  filter(!is.na(species))
nrow(d_pathogen_species)/nrow(d_pathogens)

# next we will filter instances where there is no match for species, and see if they have an exact genus match
d_pathogen_genus <- filter(d_pathogens, ! ncbi_species %in% d_pathogen_species$ncbi_species) %>%
  # join with ncbi_taxonomy where we only have one entry for each genus
  left_join(., ncbi_taxonomy %>%
              group_by(ncbi_genus) %>%
              slice_head(n = 1) %>%
              ungroup() %>%
              select(accession:species, ncbi_genus)) %>%
  filter(!is.na(genus))
nrow(d_pathogen_genus)/nrow(d_pathogens)

# next we filter any instances which did not match at the species or genus level and see if there is family match
d_pathogen_family <- filter(d_pathogens, ! ncbi_species %in% c(d_pathogen_species$ncbi_species, d_pathogen_genus$ncbi_species)) %>%
  # join with ncbi_taxonomy where we only have one entry for each family
  left_join(., ncbi_taxonomy %>%
              group_by(ncbi_family) %>%
              slice_head(n = 1) %>%
              ungroup() %>%
              select(accession:species, ncbi_family)) %>%
  filter(!is.na(family))
nrow(d_pathogen_family)/nrow(d_pathogens)

# all assigned
(nrow(d_pathogen_species) + nrow(d_pathogen_genus) + nrow(d_pathogen_family))/nrow(d_pathogens) 

# create a new pathogen list this time with accession numbers and taxonomy that links GTDB and NCBI taxonomy
d_pathogens2 <- bind_rows(d_pathogen_species, d_pathogen_genus, d_pathogen_family)
```

So ~75% of the species in our pathogen list had a perfect match to the NCBI taxonomy in the GTDB metadata file, 24% linked at the genus level and then the final few had exact matches at the Family level. Not bad.

### Plot tree

We can now plot the tree. We are plotting the tree using [ggtree](https://guangchuangyu.github.io/ggtree-book/chapter-ggtree.html) which has great documentation online.

We are going to plot the tree at the Class level (where each tip represents a unique class), colour the branches by the Phyla that contain a bacterial human pathogen, and have points at the tips that represent how many pathogen species each Class has.

```{r plot_tree, fig.width = 12, fig.height = 12}
# what phyla are in the pathogen list
d_phyla_pathogens <- group_by(d_pathogens2, ncbi_phylum) %>%
  tally()

# filter GTDB taxonomy to only keep one member of each class
taxonomy_filtered <- taxonomy %>%
  group_by(class) %>%
  slice_head(n = 1) %>%
  ungroup()

# subset tree to keep only accession numbers in the filtered taxonomy dataset
tree_sub <- keep.tip(tree, taxonomy_filtered$accession)

# make sure accessions numbers align with tip labels
tip_labels <- tibble(accession = tree_sub$tip.label) %>%
  left_join(., taxonomy_filtered)

# make tip labels class
tree_sub$tip.label <- tip_labels$class

# column for whether the phylum has a pathogenic species or not
taxonomy_filtered <- mutate(taxonomy_filtered, group = ifelse(ncbi_phylum %in% d_phyla_pathogens$ncbi_phylum, ncbi_phylum, 'other'))

# group tip labels together in terms of whether the phylum has a pathogenic species or not
# this allows for colouring of branches in the phylogenetic tree
to_group <- split(taxonomy_filtered$class, taxonomy_filtered$group)
tree_sub2 <- groupOTU(tree_sub, to_group)

# create colour palette
colour_count <- length(unique(taxonomy_filtered$group)) # number of levels
cols <- c(colorRampPalette(brewer.pal(11, "Spectral"))(nrow(d_phyla_pathogens)), 'grey')
names(cols) <- c(d_phyla_pathogens$ncbi_phylum, 'other')

# first only plot taxonomy
tree_plot <- ggtree(tree_sub2, layout = 'circular', branch.length = 'none', aes(col = group)) +
  scale_color_manual('Phylum', values = cols) +
  guides(color = guide_legend(override.aes = list(size = 5)))
unique(taxonomy_filtered$order) %>% length()

# tally number of pathogen species in each class
d_pathogens_class <- group_by(d_pathogens2, class) %>%
  tally()

# make metadata
d_meta <- tibble(class = tree_sub$tip.label) %>%
  left_join(d_pathogens_class) %>%
  mutate(n2 = replace_na(n, 0))
d_meta2 <- filter(d_meta, !is.na(n))

# create final tree plot
tree_plot2 <- tree_plot %<+% d_meta2 +
  new_scale_color() +
  geom_tippoint(aes(col = n), position = position_nudge(1), size = 4, na.rm = TRUE) +
  scale_colour_viridis_c('Number of pathogen\nspecies in each GTDB\nclass', na.value=NA) +
  geom_tiplab(aes(subset=(label %in% d_meta2$ncbi_class)), offset = 2) +
  theme(plot.margin = margin(0, 0, 0, 0, "cm"),
        legend.box.margin = margin(0, 0, 0, 2, "cm"),
        legend.text = element_text(size = 12),
        legend.title = element_text(size = 12))

tree_plot2

ggsave('Figure_1.png', tree_plot2, width = 12, height = 12)

```

### Estimating phylogenetic signal of pathogen presence

To examine any clustering in pathogens evolution across the phylogeny, we can estimate phylogenetic signal of the binary trait (tip contains pathogen: yes or no). To do this, we estimated the [delta-statistic](https://github.com/mrborges23/delta_statistic) which is specifically designed to measure phylogenetic signal in binary traits. 

As phylogenetic signal can vary at different taxonomic levels, we will calculate the delta statistic from the Order to Genus level. For each calculation, we will permute a random distribution of the delta statistic by shuffling the tips and recalculating delta 1000 times. Significant phylogenetic signal is determined to be present if the proportion of random delta calculations greater than the actual estimated delta is less than 0.05.

We will do the analyses on the Order level data first.

```{r order_signal, eval = FALSE}
# source script used for calculating the delta statistic
source('https://github.com/mrborges23/delta_statistic/raw/master/code.R')

# set number of bootstraps
n_boots = 1000

#-------------#
# 1. Order ####
#-------------#

# use an order level tree
taxonomy_order <- taxonomy %>%
  group_by(order) %>%
  # keep only one member of each order
  slice_head(n = 1) %>%
  ungroup() %>%
  select(accession, kingdom:genus)
# 1460 orders

# subset tree to only have those orders included
tree_order <- keep.tip(tree, taxonomy_order$accession)
tree_order

# add pathogen status onto this dataframe
d_pathogens_signal <- group_by(d_pathogens2, order) %>%
  tally(name = 'num_pathogens')
# 69 orders containing pathogens

# bind pathogen and taxonomy dataframes and code for pathogen, yes or no
taxonomy_order <- left_join(taxonomy_order, d_pathogens_signal) %>%
  mutate(num_pathogens = replace_na(num_pathogens, 0),
         pathogen = ifelse(num_pathogens == 0, 'no', 'yes'))
group_by(taxonomy_order, pathogen) %>%
  tally()

# reorder data frame so tip labels align with the dataframe
taxonomy_order <- left_join(tibble(accession = tree_order$tip.label), taxonomy_order) %>%
  data.frame()

# check tip labels match and node labels are unique
sum(tree_order$tip.label == taxonomy_order$accession)
# rename node labels in case they cause a random issue
tree_order$node.label <- paste('node', 1:length(tree_order$node.label), sep = '_')

# estimate delta statistic - follow recommendations from https://github.com/mrborges23/delta_statistic
pathogen_vec <- taxonomy_order$pathogen
tree_order$edge.length[tree_order$edge.length==0] <- quantile(tree_order$edge.length,0.1)*0.1
deltaA_order <- delta(pathogen_vec,tree_order,0.1,0.0589,10000,10,100)

# calculate random deltas for Order
random_delta <- tibble(boot_num = 1:n_boots)
random_delta$trait = NA

# create randomly sampled pathogen vector
for(i in 1:n_boots){
  random_delta$trait[[i]] <- data.frame(trait = sample(pathogen_vec))
}

# run this in a multisession using furrr::future_map
plan(multisession, workers = 3)

random_delta <- mutate(random_delta, output = furrr::future_map(trait, ~delta(.x$trait, tree_order, 0.1,0.0589,10000,10,10), .options = furrr_options(seed = 42)))

random_delta <- select(random_delta, boot_num, output) %>% unnest(output)

# calculate p value
p_value <- sum(random_delta$output>deltaA_order)/nrow(random_delta)

d_order_delta <- data.frame(taxonomic_level = 'order', statistic = 'delta', value = deltaA_order, p_val = sum(random_delta$output>deltaA_order)/nrow(random_delta))

d_order_delta_boot <- mutate(random_delta, taxonomic_level = 'order')
```

Next we can do the same at the Family Level.

```{r signal_family, eval = FALSE}
# use an family level tree
taxonomy_family <- taxonomy_species %>%
  group_by(family) %>%
  # keep only one member of each family
  slice_head(n = 1) %>%
  ungroup() %>%
  select(accession, kingdom:genus)
# 3650 familys

# subset tree to only have those families included
tree_family <- keep.tip(tree, taxonomy_family$accession)
tree_family

# add pathogen status onto this dataframe
d_pathogens_signal <- group_by(d_pathogens2, family) %>%
  tally(name = 'num_pathogens')
# 131 familys containing pathogens

# combine gtdb taxonomy and pathogen datasets
taxonomy_family <- left_join(taxonomy_family, d_pathogens_signal) %>%
  mutate(num_pathogens = replace_na(num_pathogens, 0),
         pathogen = ifelse(num_pathogens == 0, 'no', 'yes'))

# reorder data frame to align with tip labels
taxonomy_family <- left_join(tibble(accession = tree_family$tip.label), taxonomy_family) %>%
  data.frame()

# check tip labels match and node labels are unique
sum(tree_family$tip.label == taxonomy_family$accession)
tree_family$node.label <- paste('node', 1:length(tree_family$node.label), sep = '_')

# estimate lambda
pathogen_vec <- taxonomy_family$pathogen

# check tip labels match and node labels are unique
sum(tree_family$tip.label == taxonomy_family$accession)
tree_family$node.label <- paste('node', 1:length(tree_family$node.label), sep = '_')

# estimate delta statistic
pathogen_vec <- taxonomy_family$pathogen

tree_family$edge.length[tree_family$edge.length==0] <- quantile(tree_family$edge.length,0.1)*0.1

deltaA_family <- delta(pathogen_vec,tree_family,0.1,0.0589,10000,10,100)

# calculate random deltas for family
random_delta <- tibble(boot_num = 1:n_boots)
random_delta$trait = NA

# create randomly sampled pathogen vector
for(i in 1:n_boots){
  random_delta$trait[[i]] <- data.frame(trait = sample(pathogen_vec))
}

# parallelise calculation
plan(multisession, workers = 3)

random_delta <- mutate(random_delta, output = furrr::future_map(trait, ~delta(.x$trait, tree_family, 0.1,0.0589,10000,10,10), .options = furrr_options(seed = 42)))

random_delta <- select(random_delta, boot_num, output) %>% unnest(output)

p_value <- sum(random_delta$output>deltaA_family)/nrow(random_delta)
boxplot(random_delta$output)
abline(h = deltaA_family,col="red")

d_family_delta <- data.frame(taxonomic_level = 'family', statistic = 'delta', value = deltaA_family, p_val = sum(random_delta$output>deltaA_family)/nrow(random_delta))

d_family_delta_boot <- mutate(random_delta, taxonomic_level = 'family')

```

And finally at the Genus level.

```{r signal_genus, eval = FALSE}
#-------------#
# 3. Genus ####
#-------------#

# use an family level tree
taxonomy_genus <- taxonomy_species %>%
  group_by(genus) %>%
  # keep only one member of each family
  slice_head(n = 1) %>%
  ungroup() %>%
  select(accession, kingdom:genus)
# 15342 genera

# subset tree to only have those genera included
tree_genus <- keep.tip(tree, taxonomy_genus$accession)
tree_genus

# add pathogen status onto this dataframe
d_pathogens_signal <- group_by(d_pathogens2, genus) %>%
  tally(name = 'num_pathogens')
# 425 familys containing pathogens

taxonomy_genus <- left_join(taxonomy_genus, d_pathogens_signal) %>%
  mutate(num_pathogens = replace_na(num_pathogens, 0),
         pathogen = ifelse(num_pathogens == 0, 'no', 'yes'))
group_by(taxonomy_genus, pathogen) %>%
  tally()

# reorder data frame
taxonomy_genus <- left_join(tibble(accession = tree_genus$tip.label), taxonomy_genus) %>%
  data.frame()

# check tip labels match and node labels are unique
sum(tree_genus$tip.label == taxonomy_genus$accession)
tree_genus$node.label <- paste('node', 1:length(tree_genus$node.label), sep = '_')

# estimate lambda
pathogen_vec <- taxonomy_genus$pathogen

# check tip labels match and node labels are unique
sum(tree_genus$tip.label == taxonomy_genus$accession)
tree_genus$node.label <- paste('node', 1:length(tree_genus$node.label), sep = '_')

# estimate delta statistic
pathogen_vec <- taxonomy_genus$pathogen

tree_genus$edge.length[tree_genus$edge.length==0] <- quantile(tree_genus$edge.length,0.1)*0.1

deltaA_genus <- delta(pathogen_vec,tree_genus,0.1,0.0589,10000,10,100)

# calculate random deltas for genus
random_delta <- tibble(boot_num = 1:n_boots)
random_delta$trait = NA

# create randomly sampled pathogen vector
for(i in 1:n_boots){
  random_delta$trait[[i]] <- data.frame(trait = sample(pathogen_vec))
}

plan(multisession, workers = 3)

random_delta <- mutate(random_delta, output = furrr::future_map(trait, ~delta(.x$trait, tree_genus, 0.1,0.0589,10000,10,10), .options = furrr_options(seed = 42)))

random_delta <- select(random_delta, boot_num, output) %>% unnest(output)

p_value <- sum(random_delta$output>deltaA_genus)/nrow(random_delta)
boxplot(random_delta$output)
abline(h = deltaA_genus,col="red")

d_genus_delta <- data.frame(taxonomic_level = 'genus', statistic = 'delta', value = deltaA_genus, p_val = sum(random_delta$output>deltaA_genus)/nrow(random_delta))

d_genus_delta_boot <- mutate(random_delta, taxonomic_level = 'genus')

```

### Visualise and check phylogenetic signal

We can look at the distribution of delta statistics at each phylogenetic level and then reproduce Table 2.

```{r loading_signal, echo=FALSE}
d_order_delta <- read.csv('../data/order_delta.csv')
d_order_delta_boot <- read.csv('../data/order_delta_bootstraps.csv')

d_family_delta <- read.csv('../data/family_delta.csv')
d_family_delta_boot <- read.csv('../data/family_delta_bootstraps.csv')

d_genus_delta <- read.csv('../data/genus_delta.csv')
d_genus_delta_boot <- read.csv('../data/genus_delta_bootstraps.csv')


```

```{r plot_signal, fig.height=4, fig.width=10}
# bind dataframe together across taxonomic levels
d_signal <- bind_rows(d_order_delta, d_family_delta, d_genus_delta)
d_boots <- bind_rows(d_order_delta_boot, d_family_delta_boot, d_genus_delta_boot)

# make plot showing the random distribution of delta statistics
ggplot() +
  geom_histogram(aes(output), fill = 'light grey', col = 'black', d_boots) +
  facet_wrap(~taxonomic_level, scales = 'free') +
  theme_bw(base_size = 14) +
  geom_vline(aes(xintercept = value), col = 'red', d_signal)

# recreate Table 2.

# add columns do data
d_genus <- d_genus_delta %>%
  mutate(total_tree_tips = 15342,
         total_pathogens = 425)
d_family <- d_family_delta %>%
  mutate(total_tree_tips = 3650,
         total_pathogens = 131)
d_order <- d_order_delta %>%
  mutate(total_tree_tips = 1460,
         total_pathogens = 69)

d <- bind_rows(d_order, d_family, d_genus)

table1 <- select(d, taxonomic_level, total_tree_tips, total_pathogens, value, p_val) %>%
  mutate(., value = round(value, 2)) %>%
  gt() %>%
  cols_align(align = "center") %>%
  tab_options(table.font.name = "Arial",
              table.border.top.style = "bold",
              table.border.bottom.style = "solid",
              table.border.bottom.width = px(3),
              column_labels.border.top.color = "white",
              column_labels.border.top.width = px(3),
              column_labels.border.bottom.width = px(3),
              data_row.padding = px(10),
              heading.align = 'center')%>% 
  opt_row_striping() %>%
  cols_label(
    taxonomic_level = md("**Taxonomic level**"),
    total_tree_tips = md("**Size of GTDB tree**"),
    total_pathogens = md("**Number of tips**<br />**with pathogen presence**"),
    value = md("**delta<br />statistic**"),
    p_val = md('**p value**'))

table1

#gtsave(table1,"Table_signal.png",  path = 'plots')

```
